hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java

```diff
@@ -0,0 +1,201 @@
+    }
+
+    private ClientConfig createClientConfig() {
+        ClientConfig config = new ClientConfig();
+        ClientUserCodeDeploymentConfig clientUserCodeDeploymentConfig = new ClientUserCodeDeploymentConfig();
+        clientUserCodeDeploymentConfig.addClass("usercodedeployment.IncrementingEntryProcessor");
+        config.setUserCodeDeploymentConfig(clientUserCodeDeploymentConfig.setEnabled(true));
+        return config;
+    }
+
+    @Test
+    public void testSingleMember() {
+        ClientConfig config = createClientConfig();
+        Config clientConfig = createNodeConfig();
+
+        factory.newHazelcastInstance(clientConfig);
+        HazelcastInstance client = factory.newHazelcastClient(config);
+
+        assertCodeDeploymentWorking(client);
+    }
+
+    @Test
+    public void testWithMultipleMembers() {
+        ClientConfig config = createClientConfig();
+        Config clientConfig = createNodeConfig();
+
+        factory.newHazelcastInstance(clientConfig);
+        HazelcastInstance client = factory.newHazelcastClient(config);
+        factory.newHazelcastInstance(clientConfig);
+
+        assertCodeDeploymentWorking(client);
+    }
+
+    @Test
+    public void testWithMultipleNodes_clientReconnectsToNewNode() {
+        ClientConfig config = createClientConfig();
+        Config clientConfig = createNodeConfig();
+
+        HazelcastInstance firstInstance = factory.newHazelcastInstance(clientConfig);
+        HazelcastInstance client = factory.newHazelcastClient(config);
+        factory.newHazelcastInstance(clientConfig);
+        firstInstance.getLifecycleService().shutdown();
+
+        assertCodeDeploymentWorking(client);
+    }
+
+    private void assertCodeDeploymentWorking(HazelcastInstance client) {
+        IncrementingEntryProcessor incrementingEntryProcessor = new IncrementingEntryProcessor();
+        int keyCount = 100;
+        IMap<Integer, Integer> map = client.getMap(randomName());
+
+        for (int i = 0; i < keyCount; i++) {
+            map.put(i, 0);
+        }
+        map.executeOnEntries(incrementingEntryProcessor);
+
+        for (int i = 0; i < keyCount; i++) {
+            assertEquals(1, (int) map.get(i));
+        }
+    }
+
+    @Test
+    public void testWithInnerAndAnonymousClass() {
+        /* Implementation of IncrementingEntryProcessor insde IncrementingEntryProcessor3.jar
+
+         package usercodedeployment;
+
+         import com.hazelcast.map.AbstractEntryProcessor;
+
+         import java.util.Map;
+
+        public class IncrementingEntryProcessor extends AbstractEntryProcessor<Integer, Integer> {
+
+            @Override
+            public Object process(final Map.Entry<Integer, Integer> entry) {
+                Integer origValue = entry.getValue();
+                final Integer newValue = origValue + 1;
+
+                final usercodedeployment.IncrementingEntryProcessor.Test test = new usercodedeployment.IncrementingEntryProcessor.Test(entry);
```

Comment:
I am getting following exception here
```
java.lang.IllegalAccessError: usercodedeployment/IncrementingEntryProcessor$Test
	at usercodedeployment.IncrementingEntryProcessor.process(IncrementingEntryProcessor.java:35) ~[test-classes/:?]
	at com.hazelcast.map.impl.operation.EntryOperator.process(EntryOperator.java:328) ~[classes/:?]
	at com.hazelcast.map.impl.operation.EntryOperator.operateOnKeyValueInternal(EntryOperator.java:187) ~[classes/:?]
	at com.hazelcast.map.impl.operation.EntryOperator.operateOnKey(EntryOperator.java:172) ~[classes/:?]
	at com.hazelcast.map.impl.operation.PartitionWideEntryOperation.run(PartitionWideEntryOperation.java:84) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:194) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:120) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100) [classes/:?]
```
I verified that UserCodeDeploymentClassLoader is able to find the class definition of inner class Test and return.
But as soon as loadClass method returns it throw the exception above. 

@jerrinot any idea ?